Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp

```diff
@@ -1494,6 +1494,11 @@ void WebFrameLoaderClient::prepareForDataSourceReplacement()
     notImplemented();
 }
 
+void WebFrameLoaderClient::allowCookieAccessForWebArchive(const URL& firstParty)
+{
+    WebProcess::singleton().ensureNetworkProcessConnection().connection().send(Messages::NetworkConnectionToWebProcess::AllowCookieAccessForWebArchive(firstParty), 0);
```

Comment:
The WebProcess is not a trusted process. Why is it OK to have an IPC from the WebProcess to the network process to relax cookie access restrictions?

Reply:
This is described in the commit message.  It's only allowed if an external process has verified that a web archive has gone into the web process.  All other possible solutions involve getting the entire web archive in memory or redoing the parser, and parsing the archive outside of the web process to get the "first party" from the web archive contents.  This was the most efficient solution and it doesn't introduce new parsing into other processes.

Reply:
This line makes it safe:
NETWORK_PROCESS_MESSAGE_CHECK(m_networkProcess->webProcessDidLoadWebArchive(m_webProcessIdentifier));

Reply:
Well, "safe". It checks that the WebProcess actually loaded a WebArchive but totally blindly trusts the domain sent by the insecure WebProcess it seems.